crates/accelerate/src/commutation_checker.rs

```diff
@@ -0,0 +1,781 @@
+            CommutationLibraryEntry::QubitMapping(qm) => {
+                let out_dict = PyDict::new_bound(py);
+
+                qm.iter().for_each(|(k, v)| {
+                    out_dict
+                        .set_item(
+                            PyTuple::new_bound(
+                                py,
+                                k.iter()
+                                    .map(|q| q.map(|t| t.0))
+                                    .collect::<Vec<Option<u32>>>(),
+                            )
+                            .to_object(py),
+                            PyBool::new_bound(py, *v).to_object(py),
+                        )
+                        .ok()
+                        .unwrap()
+                });
+                out_dict.to_object(py)
+            }
+        }
+    }
+}
+
+type CacheKey = (
+    SmallVec<[Option<Qubit>; 2]>,
+    (SmallVec<[ParameterKey; 3]>, SmallVec<[ParameterKey; 3]>),
+);
+//need a struct instead of a type definition because we cannot implement serialization traits otherwise
+#[derive(Clone)]
+struct CommutationCacheEntry {
+    mapping: HashMap<CacheKey, bool>,
+}
+impl CommutationCacheEntry {
+    fn get(&self, key: &CacheKey) -> Option<&bool> {
+        self.mapping.get(key)
+    }
+    fn iter(&self) -> Iter<'_, CacheKey, bool> {
+        self.mapping.iter()
+    }
+
+    fn insert(&mut self, k: CacheKey, v: bool) -> Option<bool> {
+        self.mapping.insert(k, v)
+    }
+}
+
+impl ToPyObject for CommutationCacheEntry {
+    fn to_object(&self, py: Python) -> PyObject {
+        let out_dict = PyDict::new_bound(py);
+        for (k, v) in self.iter() {
+            let qubits = PyTuple::new_bound(
+                py,
+                k.0.iter()
+                    .map(|q| q.map(|t| t.0))
+                    .collect::<Vec<Option<u32>>>(),
+            )
+            .to_object(py);
+            let params0 =
+                PyTuple::new_bound(py, k.1 .0.iter().map(|pk| pk.0).collect::<Vec<f64>>())
+                    .to_object(py);
+            let params1 =
+                PyTuple::new_bound(py, k.1 .1.iter().map(|pk| pk.0).collect::<Vec<f64>>())
+                    .to_object(py);
+            out_dict
+                .set_item(
+                    PyTuple::new_bound(
+                        py,
+                        [
+                            qubits,
+                            PyTuple::new_bound(
+                                py,
+                                [params0, params1].iter().collect::<Vec<&PyObject>>(),
+                            )
+                            .to_object(py),
+                        ]
+                        .iter()
+                        .collect::<Vec<&PyObject>>(),
+                    ),
+                    PyBool::new_bound(py, *v).to_object(py),
+                )
+                .expect("Failed to construct commutation cache for serialization");
+        }
+        out_dict.to_object(py)
+    }
```

Comment:
Again, if you turn something into a Python native type, you don't need to call `to_object(py)` if you're sending those back to python.

If you really need a `PyObject` from a `Bound<_>`, call `.into_any()` for a `Bound<PyAny>` and then call `unbind()` which turns it into a `Py<PyAny>`.

```suggestion
let qubits = PyTuple::new_bound(
                py,
                k.0.iter()
                    .map(|q| q.map(|t| t.0))
                    .collect::<Vec<Option<u32>>>(),
            );
            let params0 =
                PyTuple::new_bound(py, k.1 .0.iter().map(|pk| pk.0).collect::<Vec<f64>>());
            let params1 =
                PyTuple::new_bound(py, k.1 .1.iter().map(|pk| pk.0).collect::<Vec<f64>>());
            out_dict
                .set_item(
                    PyTuple::new_bound(py, [qubits, PyTuple::new_bound(py, [params0, params1])]),
                    PyBool::new_bound(py, *v),
                )
                .expect("Failed to construct commutation cache for serialization");
        }
        out_dict.into_any().unbind()
```